src/main/scala/scalaz/ziokeeper/VectorClock.scala

```diff
@@ -0,0 +1,63 @@
+package scalaz.ziokeeper
+
+import scala.annotation.tailrec
+import VectorClock._
+case class VectorClock[A](value: Map[A, Long]) extends AnyVal {
+  def time(mid: A): Long = value.getOrElse(mid, 0L)
+
+  def increment(mid: A, value: Long): VectorClock[A] =
+    VectorClock(this.value.updated(mid, time(mid) + value))
+
+  def compare(right: VectorClock[A]): CompareVectorClock = {
+    val leftWithDefault  = this.value.withDefaultValue(0L)
+    val rightWithDefault = right.value.withDefaultValue(0L)
+
+    @tailrec
+    def loop(keys: List[A], leftBigger: Boolean, rightBigger: Boolean): (Boolean, Boolean) =
+      keys match {
+        case _ if leftBigger && rightBigger => (leftBigger, rightBigger)
+        case Nil                            => (leftBigger, rightBigger)
+        case h :: t =>
+          val diff = leftWithDefault(h) - rightWithDefault(h)
+          loop(t, leftBigger || diff > 0, rightBigger || diff < 0)
+      }
+
+    val keys = leftWithDefault.keySet.union(rightWithDefault.keySet)
+
+    loop(keys.toList, leftBigger = false, rightBigger = false) match {
+      case (false, false) => TheSame
+      case (true, true)   => Concurrent
+      case (true, false)  => After
+      case (false, true)  => Before
+    }
+  }
+
+  def merge(right: VectorClock[A]): VectorClock[A] = {
+    val leftWithDefault  = this.value.withDefaultValue(0L)
+    val rightWithDefault = right.value.withDefaultValue(0L)
+
+    VectorClock(
+      leftWithDefault.keySet
+        .union(rightWithDefault.keySet)
+        .foldLeft(Map.empty[A, Long]) {
+          case (acc, key) =>
+            acc + (key -> math.max(leftWithDefault(key), rightWithDefault(key)))
+        }
+    )
+  }
+}
+
+object VectorClock {
+  def empty[A] = VectorClock[A](value = Map.empty)
+
+  sealed trait CompareVectorClock
```

Comment:
This would more classically be called `Ordering` with `EQ`, `GT`, `LT`. I am not sure what `Concurrent` is.

Reply:
but this is not exactly ordering in the classic `EQ`, `GT`, `LT` sense. 

this is explanation I gave to @rabbitonweb for a question about different between `TheSame` and `Concurrent` I hope that will help you understand `Concurrent`

> the idea behind vector clock is pretty simple

> increment you local counter whatever you perform actions(e.g. add, remove, modify) as well as when you receive and send messages from/to other members of the cluster.
> when we have cluster with three member with ids 111, 222, 333 vector clock initially will look like this on all nodes.
> (111 -> 0, 222 -> 0, 333 -> 0)
> 
> at this point the are all TheSame
> 
> imagine that 111 and 222 perform add action
> 111 = (111 -> 1, 222 -> 0, 333 -> 0)
> 222 = (111 -> 0, 222 -> 1, 333 -> 0)
> 333 = (111 -> 0, 222 -> 0, 333 -> 0)
> 111 and 222 are Concurrent because you cannot say which happen first.

I need to implement Ordering/PartialOrdering to be able to sort values in ORSet so I named this differently(pretty badly though) to avoid confusion.

The problem that I struggle now is how to make sure that result will be deterministic when we have `Concurrent` update on the same key.

Reply:
renamed to `VectorClockOrdering`